### PR TITLE
[Intel HPU] enable tensor_wise_fp8

### PR DESCRIPTION
**Pull request by fmiao2372**
## Motivation
Enable tensor_wise_fp8 on HPU

> :bulb: If this PR is a Cherry Pick, the PR title needs to follow the format by adding the [Cherry-Pick] label at the very beginning and appending the original PR ID at the end. For example, [Cherry-Pick][CI] Add check trigger and logic(#5191)

> :bulb: 如若此PR是Cherry Pick，PR标题需遵循格式，在最开始加上[Cherry-Pick]标签，以及最后面加上原PR ID，例如[Cherry-Pick][CI] Add check trigger and logic(#5191) 

## Modifications
fastdeploy/model_executor/layers/backends/intel_hpu/attention/hpu_attn_backend.py
fastdeploy/model_executor/layers/backends/intel_hpu/moe/fused_moe_hpu_backend.py
fastdeploy/model_executor/layers/backends/intel_hpu/quantization/tensor_wise_fp8.py
fastdeploy/model_executor/layers/quantization/kv_cache.py
fastdeploy/model_executor/layers/quantization/tensor_wise_fp8.py
fastdeploy/worker/hpu_model_runner.py

## Usage or Command

1. set FD_HPU_MEASUREMENT_MODE=1, run BF16 with some datasets to do calibration and get activation abs-max values. 
2. offline run Model_convert.py with  activation abs-max values to get FP8 models. https://github.com/PaddlePaddle/PaddleCustomDevice/blob/develop/backends/intel_hpu/custom_ops/python/paddlenlp_ops/Model_convert.py
3. set FD_HPU_MEASUREMENT_MODE=0 and run inference.

## Checklist

- [Done] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [Done] Format your code, run `pre-commit` before commit.
- [Done] Add unit tests. Please write the reason in this PR if no unit tests.
  conducted by local tests
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -288,31 +305,41 @@ def fused_mlp_forward(self, x):
...
                 module.forward = types.MethodType(fused_self_atten_forward, module)
             if isinstance(module, Qwen2Attention):
                 module.forward = types.MethodType(fused_self_atten_forward, module)
+            if isinstance(module, Qwen3Attention):
+                module.forward = types.MethodType(fused_self_atten_forward, module)
             if isinstance(module, Ernie4_5_MLP):
+                module.measurement_mode = measurement_mode
                 module.forward = types.MethodType(fused_mlp_forward, module)
             if isinstance(module, Qwen2MLP):
                 module.forward = types.MethodType(fused_mlp_forward, module)
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/worker/hpu_model_runner.py:349]: measurement mode is only supported/verified for Ernie

File: fastdeploy/model_executor/layers/backends/intel_hpu/attention/hpu_attn_backend.py [hunk 2]
```diff
@@ -321,16 +341,38 @@ def forward_extend(
...
+                    measurement_mode=True,
+                    o_act_scale_key=o_proj_act_scale_key,
+                )
+            else:
+                out_linear_out = fused_sdpa_proj(
+                    query_states,
+                    key_value_states,
+                    forward_meta.attn_mask,
+                    None,
+                    o_proj.weight,
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/model_executor/layers/backends/intel_hpu/attention/hpu_attn_backend.py:363]: not issue

File: fastdeploy/worker/hpu_model_runner.py [hunk 3]
```diff
@@ -334,6 +361,8 @@ def __init__(
         self.device_id = device_id
         self.speculative_method = self.fd_config.speculative_config.method
         self.speculative_decoding = self.speculative_method is not None
+        # This measurement_mode only works in BF16 mode!
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/worker/hpu_model_runner.py:378]: Comment is not helpful

File: fastdeploy/model_executor/layers/backends/intel_hpu/moe/fused_moe_hpu_backend.py [hunk 4]
```diff
@@ -108,9 +138,133 @@ class HpuTensorWiseFP8MoEMethod(HpuMoEMethod):
...
+                scale_tensors = []
+                for i in logical_expert_ids:
+                    scale_tensor = get_tensor(state_dict.pop(key_template.format(i)))
+                    scale_tensors.append(scale_tensor)
+                # Stack all scale tensors into one tensor
+                stacked = paddle.stack(scale_tensors)
+                reciprocal = 1.0 / stacked
+                # Take max over all logical_expert_ids (axis=0)
+                max_tensor = paddle.min(reciprocal, axis=0)
+                return max_tensor.cast(paddle.get_default_dtype())
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/model_executor/layers/backends/intel_hpu/moe/fused_moe_hpu_backend.py]: fixed

File: fastdeploy/model_executor/layers/backends/intel_hpu/quantization/tensor_wise_fp8.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            quant_weight: The quantized weights
+            weight_scale: The scale of the quantized weights
+        """
+
+        quant_weight = get_tensor(state_dict.pop(layer.weight_key))
+        weight_scale = get_tensor(state_dict.pop(layer.weight_scale_key))
+        act_scale = get_tensor(state_dict.pop(layer.act_scale_key))
+        act_scale = 1.0 / act_scale
+
+        # quant_weight = quant_weight.transpose([1, 0]).contiguous()
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/model_executor/layers/backends/intel_hpu/quantization/tensor_wise_fp8.py]: fixed

File: fastdeploy/model_executor/forward_meta.py [hunk 6]
```diff
@@ -335,6 +335,9 @@ class HPUForwardMeta(ForwardMeta):
     #
     pre_caches_length: int = 0
 
+    #
+    measurement_mode: bool = False
```
Comment (yuanlehome) [inline fastdeploy/model_executor/forward_meta.py:346]: measurement_mode是什么意思呢？

Comment (yuanlehome) [reply fastdeploy/model_executor/forward_meta.py:346]: 可以写在代码注释里

File: fastdeploy/model_executor/layers/quantization/kv_cache.py [hunk 7]
```diff
@@ -243,6 +295,37 @@ def create_weights(self, layer: nn.Layer, **extra_weight_attrs):
                         **extra_weight_attrs,
                     },
                 )
+            if "sdpa" in layer.cache_quant_type_str:
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/quantization/kv_cache.py]: sdpa是什么意思的量化？除了这里，有文档对其进行解释嘛？

File: fastdeploy/model_executor/layers/quantization/kv_cache.py [hunk 8]
```diff
@@ -34,6 +34,7 @@ class KvCacheQuantzationTypes(str, Enum):
 
     INT8 = "int8"
     FP8 = "float8_e4m3fn"
+    FP8_E4M3 = "float8_e4m3"
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/quantization/kv_cache.py:37]: 是否考虑将kv_cache.py在hpu backend里新增一份单独维护？

File: fastdeploy/model_executor/layers/quantization/kv_cache.py (hunk 7 above)
Comment (yanfeich) [reply fastdeploy/model_executor/layers/quantization/kv_cache.py]: layer.cache_quant_type_str = "cache_fp8_sdpa_fp8" 意思是 cache 用 fp8量化，sdpa 也用fp8量化计算，也就是QK^T和 SV两个matmul是fp8的。所以除了cache 需要的k和v相关的scale，还需要q和score的相关scale.

File: fastdeploy/model_executor/layers/quantization/kv_cache.py (hunk 8 above)
Comment (yanfeich) [reply fastdeploy/model_executor/layers/quantization/kv_cache.py:37]: HPU对KV_cache的改动主要集中在：
不仅仅是对cache本身做量化，在计算SDPA的过程中就做两个低精度矩阵乘。
这导致两点变化：
1. 除了k和v需要scale，q和score同样需要scale
2. scale的数据类型是float32
另外一点不同是max_bound=240

如果觉得这些改动对目前代码比较大，别的厂家也没有类似需求，kv_cache今后也没有这方面扩展需要，那我们就把它单独拎出来，HPU单独维护一下

File: fastdeploy/model_executor/forward_meta.py (hunk 6 above)
Comment (yanfeich) [reply fastdeploy/model_executor/forward_meta.py:346]: fixed

File: fastdeploy/model_executor/layers/quantization/kv_cache.py (hunk 8 above)
Comment (yanfeich) [reply fastdeploy/model_executor/layers/quantization/kv_cache.py:37]: fixed

**Review by zoooo0820 — COMMENTED** [review]
这里能否参考 `tests/quantization`，补充下新增部分的单测；尽量保证CI覆盖率的检测

**Comment by fmiao2372** [comment]
@zoooo0820 已经添加

**Review by yuanlehome — APPROVED** [review]

**Review by EmmonsCurse — APPROVED** [review]
